lib/parallel_test_base.pm

```diff
@@ -0,0 +1,1961 @@
+            $guest_matrix{$_element}{ipaddr} = (split(/ /, get_required_var('GUEST_UNDER_TEST_IPADDR')))[$_index];
+            $guest_matrix{$_element}{nettype} = (split(/ /, get_required_var('GUEST_UNDER_TEST_NETTYPE')))[$_index];
+            $guest_matrix{$_element}{netname} = (split(/ /, get_required_var('GUEST_UNDER_TEST_NETNAME')))[$_index];
+            $guest_matrix{$_element}{netmode} = (split(/ /, get_required_var('GUEST_UNDER_TEST_NETMODE')))[$_index];
+        }
+    }
+    print "Guest Matrix After Initialization:\n", Dumper(\%guest_matrix);
+}
+
+=head2 fill_up_array
+
+Fill up or update existing array which contains information about running guest,
+mac address, ip address, network type, network name and network mode. Updated
+array will be used to set corresponding variables, for example, GUEST_UNDER_TEST_IPADDR.
+Main arguments are reference to the array, guest to be involved and variable
+related. Guest attribute name is derived from corresponding variable, for example,
+guest_matrix{guest}{netmode} is derived from GUEST_UNDER_TEST_NETMODE. At last,
+array will be filled up or updated by the latest guest_matrix{guest}{attribute}.
+=cut
+
+sub fill_up_array {
+    my ($self, %args) = @_;
+    $args{ref} //= '';
+    $args{guest} //= '';
+    $args{var} //= '';
+    croak("Array reference/guest/variable to be filled up must be given") if (!$args{ref} or !$args{guest} or !$args{var});
+
+    my @_guest_under_test = split(/ /, get_required_var('GUEST_UNDER_TEST'));
+    my $_guest_var = lc((split(/_/, $args{var}))[-1]);
+    foreach my $_guest (split(/ /, $args{guest})) {
+        if (get_var($args{var}, '')) {
+            my $_index = firstidx { $_ eq $_guest } (@_guest_under_test);
+            $args{ref}[$_index] = $guest_matrix{$_guest}{$_guest_var};
+        }
+        else {
+            push(@{$args{ref}}, $guest_matrix{$_guest}{$_guest_var});
+        }
+    }
+}
+
+=head2 test_guest_network
+
+Test networking accessibility of guest. All guests should can be reached on host
+and can reach outside from inside. The only guest that can be reached from outside
+host is the one uses host bridge network. Main arguments are guest to be tested and 
+whether die (1) or not (0) if any error. This subroutine also calls construct_uri 
+to determine the desired URI to be connected if the interested party is not localhost. 
+Please refer to subroutine construct_uri for the arguments related.
+=cut
+
+sub test_guest_network {
+    my ($self, %args) = @_;
+    $args{guest} //= '';
+    $args{die} //= 0;
+    croak("Guest to be tested must be given") if (!$args{guest});
+
+    my $_ret = 0;
+    foreach my $_guest (split(/ /, $args{guest})) {
+        record_info("Test $_guest network");
+        my $_temp = 1;
+        $_temp = script_run("timeout --kill-after=3 --signal=9 20 ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_guest ping -c5 openqa.suse.de");
```

Comment:
This won't work for O3/Tumbleweed. Suggest to use separate openqa for different product..